checkbox-provider-ce-oem/bin/check_vpu_device.py

```diff
@@ -0,0 +1,237 @@
+            "Supported VPU devices for {} is not defined".format(soc_type))
+
+    major_ver, minor_ver = kernel_version.split(".")
+    if int(major_ver) > 5 or (int(major_ver) == 5 and int(minor_ver) >= 15):
+        expected_devices.remove("ion")
+
+    return expected_devices
+
+
+def check_imx_vpu_devices():
+    """
+    Check the i.MX VPU devices
+
+    Args:
+        soc_type (str): i.MX SoC Type
+
+    Raises:
+        SystemExit: If the SoC type is not expected
+                    If VPU devices are not expected
+
+    Reference:
+        https://github.com/Freescale/libimxvpuapi#dwl-and-ewl-errors
+    """
+    result = True
+
+    expected_devices = determine_expected_imx_vpu(
+        get_soc_id(),
+        get_kernel_version()
+    )
+    nodes = [dev.name for dev in list(Path("/dev").iterdir())]
+    for dev in expected_devices:
+        if dev in nodes:
+            logging.info("The %s device is available", dev)
+        else:
+            logging.error("The %s device is not exists!!", dev)
+            result = False
+
+    if result is False:
+        raise SystemExit("# VPU devices check: Failed")
+    else:
+        logging.info("# VPU devices check: Passed")
+
+
+def get_v4l2_devices():
+    """
+    Get the name of all V4L2 devices
+
+    Returns:
+        v4l2_dev_name (list): V4L2 device name
+    """
+    v4l2_dev_name = []
+    for device in sorted(
+            Path("/sys/class/video4linux").glob("video*/name")):
+        v4l2_dev_name.append(device.read_text().strip("\n"))
+
+    return v4l2_dev_name
+
+
+def check_mtk_vpu_devices():
+    """
+    Detect MTK Video processing unit is available
+
+    Raises:
+        SystemExit: if no VPU device is detected or
+                    the device node is not expected
+
+    Reference:
+        https://mediatek.gitlab.io/aiot/doc/aiot-dev-guide/master/sw/yocto/release-notes/aiot-yocto-v22.2-release-note.html
+    """
+    expected_device_pattern = [
+        "mtk-vcodec-dec", "mtk-vcodec-enc", "mtk-mdp[0-9]*:m2m"]
+
+    check_result = []
+    for pattern in expected_device_pattern:
+        for dev_name in get_v4l2_devices():
+            if re.search(pattern, dev_name):
+                check_result.append(dev_name)
+                logging.info("VPU %s device detected", dev_name)
+                break
+
+    if len(check_result) == len(expected_device_pattern):
+        logging.info("# VPU devices check: Passed")
+    else:
+        raise SystemExit("# VPU devices check: Failed")
+
+
+def main():
+    """
+    Detect Video processing unit is available
+
+    Raises:
+        SystemExit: if no VPU device is detected or
+                    the device node is not expected
+    """
+    init_logger()
+    supported_funcions = {
+        "Freescale i.MX": check_imx_vpu_devices,
+        "jep106:0426": check_mtk_vpu_devices
+    }
+
+    family = get_soc_family()
+    if family in supported_funcions.keys():
+        supported_funcions.get(family)()
```

Comment:
Why are there 2 brackets `()` here?

Reply:
it's because the items in supported_functions dictionary is function

Reply:
Oh. I didn't notice that. Thanks for the explaination.